### PR TITLE
Don't always use object as the type for failed unary operator

### PR DESCRIPTION
**Pull request by jcouv**
Fixes https://github.com/dotnet/roslyn/issues/21486

### CODE REVIEW

**Comment by jcouv** [comment]
@dotnet/roslyn-compiler for review. Thanks

File: src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs [hunk 1]
```diff
@@ -2233,7 +2233,7 @@ private BoundExpression BindUnaryOperatorCore(CSharpSyntaxNode node, string oper
                 return new BoundUnaryOperator(node, kind, operand, ConstantValue.NotAvailable,
                     methodOpt: null,
                     resultKind: LookupResultKind.Empty,
-                    type: GetSpecialType(SpecialType.System_Object, diagnostics, node),
+                    type: operand.Type ?? GetSpecialType(SpecialType.System_Object, diagnostics, node),
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs]: We should be using an error type here. That is exactly what error types are for.

**Comment by jcouv** [comment]
test windows_debug_unit32_prtest  please

**Comment by jcouv** [comment]
windows_debug_unit32_prtest failed on CompilerServerUnitTests.CscFallBackOutputNoUtf8 ([details](https://ci.dot.net/job/dotnet_roslyn/job/post-dev15.5-contrib/job/windows_debug_unit32_prtest/56/))

@dotnet/roslyn-infrastructure Please advise.

**Comment by jasonmalinowski** [comment]
@jcouv It's not a known failure. Feel free to resume. If you're seeing it multiple times then I'd suspect your change is causing it somehow.

**Comment by jcouv** [comment]
@jasonmalinowski I don't know that this occurs "multiple times". I think it's just the caching that kicks in.

**Comment by jcouv** [comment]
@jasonmalinowski Confirmed this was a spurious error.

**Comment by AlekseyTs** [comment]
>                     GetSpecialType(SpecialType.System_Object, diagnostics, node),

It feels like this place should be fixed as well.

---
Refers to: src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs:2273 in c202820. [](commit_id = c202820051d3121bba2a972f5dc6f89598fb3ef4, deletion_comment = False)

**Comment by AlekseyTs** [comment]
>                     GetSpecialType(SpecialType.System_Object, diagnostics, node),

It feels like this place should be fixed as well.

---
Refers to: src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs:1923 in c202820. [](commit_id = c202820051d3121bba2a972f5dc6f89598fb3ef4, deletion_comment = False)

**Comment by AlekseyTs** [comment]
>                     GetSpecialType(SpecialType.System_Object, diagnostics, node),

It feels like this place should be fixed as well.

---
Refers to: src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs:1961 in c202820. [](commit_id = c202820051d3121bba2a972f5dc6f89598fb3ef4, deletion_comment = False)

**Pull request merged**